### PR TITLE
work-in-progress: add support for organization management

### LINKED ISSUE #1719: Feature Request: UI for adding organizations to my Spoke instance
**Problem**
Many orgs need to add organizations to their Spoke instances and have historically used the `SUPPRESS_SELF_INVITE` env var toggle method or directly edited the orgs into the database. This is pretty challenging for a non-technical Spoke admin and makes it hard to make changes to organizations without tech support.

**Solution**
It would be awesome to have a screen that only a user who created an instance could access to add organizations to the instance. My initial thought for the design of this would be to have a menu item in the upper right account menu under a new section that you'd have to create called "admin tools":
Screen Shot 2020-07-29 at 3 22 55 PM
and clicking the "manage organizations" link would take the admin to a list of their orgs that would look similar to the campaigns screen. The MVP is to be able to add orgs and see this screen if an environment variable called MULTI_TENANT (that you'd also need to add) is set to true.

### PR DESCRIPTION
**Pull request by wesmoncrief**
# Fixes #1719

## Description

WIP - here's what we have so far to create/manage organizations. Looking for intermediate feedback!

# Checklist:

- [ ] I have manually tested my changes on desktop and mobile
- [ ] The test suite passes locally with my changes
- [ ] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [ ] [My change is 300 lines of code or less](https://github.com/MoveOnOrg/Spoke/blob/main/CONTRIBUTING.md#your-first-code-contribution), or has a documented reason in the description why it’s longer
- [ ] I have made any necessary changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] My PR is labeled [WIP] if it is in progress
Spoke_multi_tenant1

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -49,4 +49,5 @@ TWILIO_MESSAGE_VALIDITY_PERIOD=
 DST_REFERENCE_TIMEZONE='America/New_York'
 PASSPORT_STRATEGY=local
 EXPERIMENTAL_TAGS=1
-TEXTER_SIDEBOXES=contact-reference,tag-contact
\ No newline at end of file
+TEXTER_SIDEBOXES=contact-reference,tag-contact
+MULTI_TENANT=true
```
Comment (schuyler1d) [inline .env.example]: Every Spoke instance is multi-tenant, so I don't think we need a variable for this.  Please remove.

File: __test__/containers/UserMenu.test.js [hunk 2]
```diff
@@ -8,35 +8,43 @@ import { StyleSheetTestUtils } from "aphrodite";
 import { UserMenu } from "../../src/containers/UserMenu";
 import MuiThemeProvider from "material-ui/styles/MuiThemeProvider";
 
+function getData(isSuperAdmin = "false") {
+  return {
+    currentUser: {
+      id: 1,
+      displayName: "TestName",
+      email: "test@test.com",
+      is_superadmin: isSuperAdmin,
```
Comment (schuyler1d) [inline __test__/containers/UserMenu.test.js:17]: is_superadmin should probably be `isSuperAdmin` here to keep to the camelCase norm for graphql

File: __test__/containers/UserMenu.test.js [hunk 3]
```diff
@@ -8,35 +8,43 @@ import { StyleSheetTestUtils } from "aphrodite";
 import { UserMenu } from "../../src/containers/UserMenu";
 import MuiThemeProvider from "material-ui/styles/MuiThemeProvider";
 
+function getData(isSuperAdmin = "false") {
```
Comment (schuyler1d) [inline __test__/containers/UserMenu.test.js]: isSuperAdmin should be a boolean, not a string, so remove the parens?

File: src/api/user.js [hunk 4]
```diff
@@ -7,6 +7,7 @@ export const schema = `
     displayName: String
     email: String
     cell: String
+    is_superadmin: Boolean
```
Comment (schuyler1d) [inline src/api/user.js:10]: isSuperAdmin

File: src/routes.jsx [hunk 5]
```diff
@@ -175,13 +177,22 @@ export default function makeRoutes(requireAuth = () => {}) {
         </Route>
       </Route>
       <Route path="login" component={Login} />
+      <Route
+        path="adminOrganizations"
```
Comment (schuyler1d) [inline src/routes.jsx]: maybe just "organizations" so we don't need caps in the paths

File: src/containers/CreateAdditionalOrganization.jsx [hunk 6]
```diff
@@ -0,0 +1,169 @@
+import PropTypes from "prop-types";
```
Comment (schuyler1d) [inline src/containers/CreateAdditionalOrganization.jsx:1]: In most of the codebase, we would make this a component and then create a dialog in the AdminOrganizationDashboard.  I suppose this makes it possible to get a link directly to adding an organization (e.g. on first startup)?

File: src/containers/UserMenu.jsx [hunk 7]
```diff
@@ -121,6 +141,11 @@ export class UserMenu extends Component {
               {currentUser.email}
             </MenuItem>
             <Divider />
+            {window.MULTI_TENANT && isSuperAdmin ? (
```
Comment (schuyler1d) [inline src/containers/UserMenu.jsx]: remove `window.MULTI_TENANT &&` -- anyone that's a superadmin can see this -- we don't need to turn it on.

File: src/server/api/organization.js [hunk 8]
```diff
@@ -21,8 +21,16 @@ export const resolvers = {
       await accessRequired(user, organization.id, "SUPERVOLUNTEER");
       return getCampaigns(organization.id, cursor, campaignsFilter, sortBy);
     },
+    campaignsCount: async (organization, _, { user }) => {
+      await accessRequired(user, organization.id, "OWNER", true);
+      return getCampaignsCount(organization.id);
```
Comment (schuyler1d) [inline src/server/api/organization.js]: `return r.getCount(r.knex("campaign").where({"organization_id": organization.id, is_archived: false}))` is simpler than the indirection with buildCampaignQuery, but if you have use cases that will grow then this is fine.

**Review by schuyler1d — CHANGES_REQUESTED** [review]
This is a really important interface -- thank you!

most feedback are nits -- the one thing I'd like is to remove the MULTI_TENANT variable/conditions.

File: src/server/api/organization.js (hunk 8 above)
Comment (wesmoncrief) [reply src/server/api/organization.js]: Yeah that seems better - I'll update!

File: src/containers/UserMenu.jsx (hunk 7 above)
Comment (wesmoncrief) [reply src/containers/UserMenu.jsx]: cool!

File: src/containers/CreateAdditionalOrganization.jsx (hunk 6 above)
Comment (wesmoncrief) [reply src/containers/CreateAdditionalOrganization.jsx:1]: Do you mean a [material UI dialog](https://material-ui.com/components/dialogs/)? 
Yeah I think we kinda just added this to `routes` as a first step as we were getting things set up, so I don't think any of us are tied to this being a route.
I'd be happy to change if you or someone else thinks it's worth it
cc @ibrand  @lesia @ali @kelwen

File: src/routes.jsx (hunk 5 above)
Comment (wesmoncrief) [reply src/routes.jsx]: sure

File: __test__/containers/UserMenu.test.js (hunk 3 above)
Comment (wesmoncrief) [reply __test__/containers/UserMenu.test.js]: yep!

File: src/api/user.js (hunk 4 above)
Comment (wesmoncrief) [reply src/api/user.js:10]: done

File: __test__/containers/UserMenu.test.js (hunk 2 above)
Comment (wesmoncrief) [reply __test__/containers/UserMenu.test.js:17]: done

**Comment by schuyler1d** [comment]
I changed your PR to target `stage-main-81fast` branch which is our next release target.  Can pull against that and resolve conflicts?  would love to get this into 8.1

**Comment by wesmoncrief** [comment]
Yeah I can do that! There's a couple things I'm going to fix up this
evening, I'll let ya know when it's done

Wes

File: .env.example (hunk 1 above)
Comment (wesmoncrief) [reply .env.example]: done!

**Pull request merged**